Index: mtools-4.0.10/charsetConv.c
===================================================================
--- mtools-4.0.10.orig/charsetConv.c	2009-11-13 10:48:25.013503707 -0500
+++ mtools-4.0.10/charsetConv.c	2009-11-13 10:48:26.223493795 -0500
@@ -410,7 +410,7 @@
 			break;
 		native += r;
 	}
-	if(mangled && end && native < end)
+	if(mangled && ((end && native < end) || (!end && *native &&  i == len)))
 		*mangled |= 3;
 	wchar[i]='\0';
 	return i;
Index: mtools-4.0.10/file_name.c
===================================================================
--- mtools-4.0.10.orig/file_name.c	2009-11-13 10:54:47.843508064 -0500
+++ mtools-4.0.10/file_name.c	2009-11-13 10:55:55.412317381 -0500
@@ -63,10 +63,6 @@
 
 	*Case = NONE;
 	for( ;  *s ; s++) {
-		if(!count) {
-			*mangled |= 3;
-			break;
-		}
 		/* skip spaces & dots */
 		if(*s == ' ' || *s == '.') {
 			*mangled |= 3;
@@ -91,7 +87,6 @@
 				*Case = UPPER;
 		} else
 			*t = *s;
-		count--;
 		t++;
 	}
 	wchar_to_dos(toDos, buffer, out, t - buffer, mangled);
