Index: mtools-4.0.10/scripts/uz
===================================================================
--- mtools-4.0.10.orig/scripts/uz	2010-04-13 22:55:33.000000000 +0200
+++ mtools-4.0.10/scripts/uz	2010-04-13 23:02:34.000000000 +0200
@@ -42,10 +42,12 @@
 case $0 in
 *uz)
 	tarparam="-pxvf"
+	zipparam=""
 	action="Extracting from "
 	;;
 *lz)
 	tarparam="-tvf"
+	zipparam="-l"
 	action="Reading directory of "
 	;;
 *)
@@ -64,23 +66,31 @@
 	echo >&2
 	found=
 
-	for suffix in "" .gz .tgz .tar.gz .z .tar.z .taz .tpz .Z .tar.Z .tar.bz2; do
+	for suffix in "" .gz .tgz .tar.gz .z .tar.z .taz .tpz .Z .tar.Z .tar.bz2 .zip; do
 		if [ -r "${1}$suffix" ]; then
 			found=$1$suffix
 			break
 		fi
 	done
 
+	unzip=0
 	case $found in
 		*.tar.bz2 | *.tb2)
 			uzcmd='bzip2 -cd'
 			;;
+		*.zip)
+			unzip=1
+			;;
 	esac
 	if [ -z "$found" ]; then
 		echo "$0: could not read \"$1\"." >&2
 	else
 		echo "$action \"$found\"." >&2
-		$uzcmd -- "$found" | tar "$tarparam" -
+		if [ $unzip = 1 ]; then
+			unzip $zipparam -- "$found"
+		else
+			$uzcmd -- "$found" | tar "$tarparam" -
+		fi
 	fi
 	shift
 done
